csharp/ql/src/semmle/code/csharp/Stmt.qll

```diff
@@ -582,6 +582,27 @@ class ForeachStmt extends LoopStmt, @foreach_stmt {
    */
   Expr getIterableExpr() { result = this.getChild(1) }
 
+  /** Gets the called `GetEnumerator` method. */
+  Method getGetEnumerator() { foreach_stmt_desugar(this, result, 1) }
+
+  /** Gets the called `MoveNext` or `MoveNextAsync` method. */
+  Method getMoveNext() { foreach_stmt_desugar(this, result, 3) }
+
+  /** Gets the called `Dispose` or `DisposeAsync` method. */
+  Method getDispose() { foreach_stmt_desugar(this, result, 4) }
+
+  /** Gets the called `Current` property. */
+  Property getCurrent() { foreach_stmt_desugar(this, result, 2) }
+
+  /**
+   * Gets the intermediate type to which the `Current` property is converted before
+   * being converted to the iteration variable type.
+   */
+  Type getElementType() { foreach_stmt_desugar(this, result, 5) }
```

Comment:
How does this differ from `this.getCurrent().getType()` and `this.getVariable(0).getType()`?

Reply:
In most cases it will match both of those. And in the following case it will not match any:
```
var c = new A();
foreach(E e in c) { }

public class A
{
    public B GetEnumerator() => throw null;
}

public class B
{
    public C Current { get; }
    public bool MoveNext() => throw null;
}

public class C
{
    public static implicit operator D(C c)=> throw null;
}

public class D { }
public class E : D { }
```
The above is converted to:
```
private static void <Main>$(string[] args)
{
    A a = new A();
    B enumerator = a.GetEnumerator();
    try
    {
        while (enumerator.MoveNext())
        {
            E e = (E)(D)enumerator.Current;
        }
    }
    finally
    {
        IDisposable disposable = enumerator as IDisposable;
        if (disposable != null)
        {
            disposable.Dispose();
        }
    }
}
```
Note that `Current` has type `C`, the variable has type `E`, and there's an extra cast to `D`, which is the element type. It is described here: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#the-foreach-statement

Reply:
Thanks for the detailed explanation.